### ISSUE TITLE
openpgp-tool gen-key fails on Yubikey 4 due to hard coded public exponent length

### ISSUE THREAD

**Issue by jdh--** [OpenSC/OpenSC#1949]
### Problem Description
`openpgp-tool --gen-key` fails when run on yubikey 4's openpgp app due to setting the exponent length to 32bits instead of it's supported 17bit

When doing a PUT DATA on Algorithm attributes, the `Length of public exponent e in bits` is hard coded to 0x20 (32bits), but the openpgp card app on my example yubikey 4 only accepts 0x11 (17bits)

Why yubico did this in their openpgp card implementation, I do not know.

changing `key-attr` through `gpg --card-edit` works on this yubikey because the scdaemon implementation copies the current `algorithm attributes` by reading them from the `Application Related Data` tag data object, and only changes the `Length of modulus n in bits` leaving the existing exponent value at 0x11

This [gnupg-devel thread](https://lists.gnupg.org/pipermail/gnupg-devel/2018-July/033835.html)  discusses the ambiguity in the spec, and issues when switching algorithms without known defaults

For newer devices, openpgp card 3.4 adds the `Algorithm Information` data object which could be used to lookup supported values when setting `algorithm attributes`

### Proposed Resolution
One option is to mimic [scdaemon's scd/app-openpgp.c](https://github.com/gpg/gnupg/blob/d272308cc52c90a3a7c54fa542fbf871a0431224/scd/app-openpgp.c#L3560) implementation in change_rsa_keyattr which reads the existing values and only changes modulus length, preserving the previous exponent value.
Corner cases when switching algorithms still have hard coded defaults which will fail on yubikey 4s: gpg/gnupg@29692718768c28c524be6306081ab1852e75fe07

Alternately, allow the RSA public exponent length value to be set from the command line via something like `--rsa-e-len 17` with the default being 32bits, or bake the setting into an extended key type as in `--key-type rsa4096e17`

For openpgp cards >= 3.4 consider implementing logic to validate algorithm attributes against supported algorithms and values from the `Algorithm Information` data object

### Steps to reproduce
With a Yubikey 4 running OpenPGP Card 2.1 app:

`openpgp-tool --verify CHV3 --pin 12345678 --gen-key 3 --key-type rsa3072`
it will error with "error: failed to generate key: Incorrect parameters in APDU"

Using `opensc-tool` you can send the CHV3 unlock (with default admin PIN) and then send the PUT DATA command that `openpgp-tool` is sending with 00:20 as the exponent length and see the 0x6A 0x80 "Incorrect parameters in the command data field" response

Run again with 00:11 as the exponent length, you get the 0x90 "Command correct" response

```
$ opensc-tool -s 00:20:00:83:08:31:32:33:34:35:36:37:38 -s 00:DA:00:C3:06:01:0C:00:00:20:00
Using reader with a card: Yubico Yubikey 4 OTP+U2F+CCID
Sending: 00 20 00 83 08 31 32 33 34 35 36 37 38 
Received (SW1=0x90, SW2=0x00)
Sending: 00 DA 00 C3 06 01 0C 00 00 20 00 
Received (SW1=0x6A, SW2=0x80)
$ opensc-tool -s 00:20:00:83:08:31:32:33:34:35:36:37:38 -s 00:DA:00:C3:06:01:0C:00:00:11:00
Using reader with a card: Yubico Yubikey 4 OTP+U2F+CCID
Sending: 00 20 00 83 08 31 32 33 34 35 36 37 38 
Received (SW1=0x90, SW2=0x00)
Sending: 00 DA 00 C3 06 01 0C 00 00 11 00 
Received (SW1=0x90, SW2=0x00)
```

### Logs
```
$ openpgp-tool --verify CHV3 --pin 12345678 --gen-key 3 --key-type rsa3072
Using reader with a card: Yubico Yubikey 4 OTP+U2F+CCID
error: failed to generate key: Incorrect parameters in APDU
```
Debug output on where you can see the 0x20 exponent length sent and failed for APDU command "00 DA 00 C3 06 01 0C 00 00 20 00"
```
P:23187; T:0x140735582913408 01:42:11.117 [openpgp-tool] apdu.c:363:sc_single_transmit: called
P:23187; T:0x140735582913408 01:42:11.117 [openpgp-tool] apdu.c:370:sc_single_transmit: CLA:0, INS:DA, P1:0, P2:C3, data(6) 0x7f84f4c00140
P:23187; T:0x140735582913408 01:42:11.117 [openpgp-tool] reader-pcsc.c:297:pcsc_transmit: reader 'Yubico Yubikey 4 OTP+U2F+CCID'
P:23187; T:0x140735582913408 01:42:11.117 [openpgp-tool] reader-pcsc.c:298:pcsc_transmit: 
Outgoing APDU (11 bytes):
00 DA 00 C3 06 01 0C 00 00 20 00 ......... .
P:23187; T:0x140735582913408 01:42:11.117 [openpgp-tool] reader-pcsc.c:216:pcsc_internal_transmit: called
P:23187; T:0x140735582913408 01:42:11.118 [openpgp-tool] reader-pcsc.c:307:pcsc_transmit: 
Incoming APDU (2 bytes):
6A 80 j.
P:23187; T:0x140735582913408 01:42:11.118 [openpgp-tool] apdu.c:382:sc_single_transmit: returning with: 0 (Success)
P:23187; T:0x140735582913408 01:42:11.118 [openpgp-tool] apdu.c:535:sc_transmit: returning with: 0 (Success)
P:23187; T:0x140735582913408 01:42:11.118 [openpgp-tool] card.c:523:sc_unlock: called
P:23187; T:0x140735582913408 01:42:11.118 [openpgp-tool] iso7816.c:128:iso7816_check_sw: Incorrect parameters in the data field
P:23187; T:0x140735582913408 01:42:11.118 [openpgp-tool] card-openpgp.c:1829:pgp_put_data_plain: Card returned error: -1205 (Incorrect parameters in APDU)
P:23187; T:0x140735582913408 01:42:11.118 [openpgp-tool] card-openpgp.c:1891:pgp_put_data: PUT DATA returned error: -1205 (Incorrect parameters in APDU)
P:23187; T:0x140735582913408 01:42:11.118 [openpgp-tool] card-openpgp.c:2345:pgp_update_new_algo_attr: Cannot set new algorithm attributes: -1205 (Incorrect parameters in APDU)
P:23187; T:0x140735582913408 01:42:11.118 [openpgp-tool] card-openpgp.c:2817:pgp_gen_key: Cannot set attributes for new-generated key: -1205 (Incorrect parameters in APDU)
P:23187; T:0x140735582913408 01:42:11.118 [openpgp-tool] card-openpgp.c:3348:pgp_card_ctl: returning with: -1205 (Incorrect parameters in APDU)
P:23187; T:0x140735582913408 01:42:11.119 [openpgp-tool] card.c:1024:sc_card_ctl: returning with: -1205 (Incorrect parameters in APDU)
error: failed to generate key: Incorrect parameters in APDU
P:23187; T:0x140735582913408 01:42:11.119 [openpgp-tool] card.c:523:sc_unlock: called
P:23187; T:0x140735582913408 01:42:11.119 [openpgp-tool] reader-pcsc.c:709:pcsc_unlock: called
P:23187; T:0x140735582913408 01:42:11.119 [openpgp-tool] card.c:414:sc_disconnect_card: called
P:23187; T:0x140735582913408 01:42:11.119 [openpgp-tool] reader-pcsc.c:642:pcsc_disconnect: Yubico Yubikey 4 OTP+U2F+CCID:SCardDisconnect returned: 0x00000000
P:23187; T:0x140735582913408 01:42:11.120 [openpgp-tool] card.c:436:sc_disconnect_card: returning with: 0 (Success)
P:23187; T:0x140735582913408 01:42:11.120 [openpgp-tool] ctx.c:927:sc_release_context: called
P:23187; T:0x140735582913408 01:42:11.120 [openpgp-tool] reader-pcsc.c:946:pcsc_finish: called
```

Successfully Running against an old Crypto Stick v1.2 with OpenPGP Card 2.0
```
$ openpgp-tool --verify CHV3 --pin 12345678 --gen-key 3 --key-type rsa3072
Using reader with a card: German Privacy Foundation Crypto Stick v1.2
Fingerprint:
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX XXX
```
debug output shows the 0x20 exponent length sent and accepted by the cryptostick
```
P:23267; T:0x140735582913408 01:52:35.990 [openpgp-tool] apdu.c:363:sc_single_transmit: called
P:23267; T:0x140735582913408 01:52:35.990 [openpgp-tool] apdu.c:370:sc_single_transmit: CLA:0, INS:DA, P1:0, P2:C3, data(6) 0x7fb3c6e00850
P:23267; T:0x140735582913408 01:52:35.990 [openpgp-tool] reader-pcsc.c:297:pcsc_transmit: reader 'German Privacy Foundation Crypto Stick v1.2'
P:23267; T:0x140735582913408 01:52:35.990 [openpgp-tool] reader-pcsc.c:298:pcsc_transmit: 
Outgoing APDU (11 bytes):
00 DA 00 C3 06 01 0C 00 00 20 00 ......... .
P:23267; T:0x140735582913408 01:52:35.990 [openpgp-tool] reader-pcsc.c:216:pcsc_internal_transmit: called
P:23267; T:0x140735582913408 01:52:36.040 [openpgp-tool] reader-pcsc.c:307:pcsc_transmit: 
Incoming APDU (2 bytes):
90 00 ..
P:23267; T:0x140735582913408 01:52:36.041 [openpgp-tool] apdu.c:382:sc_single_transmit: returning with: 0 (Success)
P:23267; T:0x140735582913408 01:52:36.041 [openpgp-tool] apdu.c:535:sc_transmit: returning with: 0 (Success)
P:23267; T:0x140735582913408 01:52:36.041 [openpgp-tool] card.c:523:sc_unlock: called
```

**Comment by dengert** [comment]
Other options:
- card-openpgp.c can determine its a Yubikey by the ATR. It already does this. It could then then send 0x11 as length. 

- If length of 0x20 fails, try the APDU again with length 0x11

**Comment by marschap** [comment]
@jdj Please have a look at PR #1959
It uses an approach very similar to what gpg does.

Can you try if it fixes the issue?
You can download the PR as patch from [here](https://patch-diff.githubusercontent.com/raw/OpenSC/OpenSC/pull/1959.patch)

**Comment by marschap** [comment]
This issue should be fixed by merged PR #1959

**Comment by Jakuje** [comment]
Closing as fixed with the PR from previous comment.
